File: frontend/src/lib/time.ts
// ============================================================
// Experiment Controller — Time Helpers
// WIB (Asia/Jakarta, UTC+7) formatting and offset calculation
// ============================================================

const WIB_TIMEZONE = "Asia/Jakarta";

/**
 * Format a Date (or timestamp) in WIB with millisecond precision.
 * Output: "YYYY-MM-DD HH:MM:SS.sss"
 */
export function formatWIB(date: Date | number): string {
  const d = typeof date === "number" ? new Date(date) : date;

  const formatter = new Intl.DateTimeFormat("en-CA", {
    timeZone: WIB_TIMEZONE,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
  });

  const parts = formatter.formatToParts(d);
  const get = (type: Intl.DateTimeFormatPartTypes) =>
    parts.find((p) => p.type === type)?.value ?? "00";

  const ms = d.getMilliseconds().toString().padStart(3, "0");

  return `${get("year")}-${get("month")}-${get("day")} ${get("hour")}:${get("minute")}:${get("second")}.${ms}`;
}

/**
 * Get the current time as Unix epoch milliseconds.
 */
export function nowMs(): number {
  return Date.now();
}

/**
 * Calculate client offset from server time.
 * offset = server_time_ms - client_time_ms
 * Positive offset means client is behind server.
 */
export function calculateOffset(serverTimeWib: string, clientTimeMs: number): number {
  const serverMs = new Date(serverTimeWib).getTime();
  return serverMs - clientTimeMs;
}

/**
 * Format milliseconds as MM:SS display for countdowns.
 */
export function formatCountdown(ms: number): string {
  if (ms <= 0) return "00:00";
  const totalSeconds = Math.ceil(ms / 1000);
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  return `${minutes.toString().padStart(2, "0")}:${seconds.toString().padStart(2, "0")}`;
}

/**
 * Format milliseconds as MM:SS.s display with tenths.
 */
export function formatCountdownPrecise(ms: number): string {
  if (ms <= 0) return "00:00.0";
  const totalTenths = Math.ceil(ms / 100);
  const minutes = Math.floor(totalTenths / 600);
  const seconds = Math.floor((totalTenths % 600) / 10);
  const tenths = totalTenths % 10;
  return `${minutes.toString().padStart(2, "0")}:${seconds.toString().padStart(2, "0")}.${tenths}`;
}

File: backend/internal/wib/time.go
// ============================================================
// WIB — Western Indonesian Time helpers
// ============================================================
package wib

import (
	"fmt"
	"time"
)

var jakartaLoc *time.Location

func init() {
	var err error
	jakartaLoc, err = time.LoadLocation("Asia/Jakarta")
	if err != nil {
		panic(fmt.Sprintf("failed to load Asia/Jakarta timezone: %v", err))
	}
}

// Now returns current time in WIB.
func Now() time.Time {
	return time.Now().In(jakartaLoc)
}

// Format formats a time in WIB with millisecond precision.
// Output: "2006-01-02 15:04:05.000"
func Format(t time.Time) string {
	return t.In(jakartaLoc).Format("2006-01-02 15:04:05.000")
}

// FormatISO formats a time in WIB as ISO 8601 with timezone.
// Output: "2006-01-02T15:04:05.000+07:00"
func FormatISO(t time.Time) string {
	return t.In(jakartaLoc).Format("2006-01-02T15:04:05.000+07:00")
}

// Location returns the Asia/Jakarta location.
func Location() *time.Location {
	return jakartaLoc
}

// UnixMsToTime converts Unix milliseconds to time.Time.
func UnixMsToTime(ms int64) time.Time {
	return time.UnixMilli(ms).In(jakartaLoc)
}

// TimeToUnixMs converts time.Time to Unix milliseconds.
func TimeToUnixMs(t time.Time) int64 {
	return t.UnixMilli()
}

// ComputeOffset computes the offset between server and client time in ms.
// offset = server_ms - client_ms
func ComputeOffset(serverTime time.Time, clientTimeMs int64) int64 {
	return serverTime.UnixMilli() - clientTimeMs
}
